util/gen_action_info.py

```diff
@@ -0,0 +1,250 @@
+        "crafting": "crafting_action_info.js",
+        "combo": "pve_action_combos.js",
+        "invalid": "invalid_action.log",
+    },
+    "locale_url": {
+        "root": "https://raw.githubusercontent.com/",
+        "intl": "xivapi/ffxiv-datamining/master/csv/",
+        "cn": "thewakingsands/ffxiv-datamining-cn/master/",
+        "ko": "Ra-Workspace/ffxiv-datamining-ko/master/csv/",
+        "local": "",
+    },
+    "path": {"cactbot": os.path.abspath(__file__)[:-24]},
+    "log": {"error": "gen_action_info.log"},
+}
+
+
+def tree():
+    return defaultdict(tree)
+
+
+def __get_remote_table(url, inputs, outputs=None):
+    """Connects to a remote source to retrieve the table data"""
+    # TODO: Error handling
+    with urllib.request.urlopen(url) as response:
+        csv_file = csv.reader(io.StringIO(response.read().decode("utf-8-sig")))
+
+    # First line is the indices and third line is the data types.  They aren't currently used, so we discard them.
+    next(csv_file)
+    # Second line is the headers
+    # Append the hexidecimal version of the ID as a new column so the original ID is usable for correllation
+    headers = next(csv_file) + ["HexID"]
+    next(csv_file)
+
+    # Change # to ID for more readable data
+    headers[0] = headers[0].replace("#", "ID")
+    # Generate the hex ID from the ID and return the data with the headers prepended
+    return [headers] + [x + [format(int(x[0]), "X")] for x in csv_file]
+
+
+def __get_local_table(filename, inputs, outputs=None):
+    """Gets table data from a local file"""
+    # I'm honestly not even sure if this will work as expected. It can be updated/fixed if a use case is found.
+    with open(filename, "r") as table:
+        csv_file = [x.rstrip("\r\n").split(",") for x in table.read().decode("utf-8-sig")]
+
+    # First line is the indices and third line is the data types.  They aren't currently used, so we discard them.
+    csv_file.pop(2)
+    csv_file.pop(0)
+
+    # Next line is the headers
+    # Append the hexidecimal version of the ID as a new column so the original ID is usable for correllation
+    headers = csv_file.pop(0) + ["HexID"]
+    # Change # to ID for more readable data
+    headers[0] = headers[0].replace("#", "ID")
+    # Generate the hex ID from the ID and return the data with the headers prepended
+    return [headers] + [x + [format(int(x[0]), "X")] for x in csv_file]
+
+
+def get_data_table(table_name, locale="intl", inputs=None, outputs=None):
+    """Retrieves table data based on provided locale"""
+    if locale == "local" and os.path.exists(table_name):
+        return __get_local_table(table_name + ".csv", inputs, outputs)
+    if locale in config["locale_url"]:
+        url = config["locale_url"]["root"] + config["locale_url"][locale] + table_name + ".csv"
+        return __get_remote_table(url, inputs, outputs)
+    else:
+        raise Exception("Invalid locale: %s" % locale)
+
+
+def normalize_name(str):
```

Comment:
Can you use the `csv_util` function here?

Reply:
I originally pulled the capability from csv_util, but I don't like the logic of how it does it.  I think it is too rigid to handle alternative use cases.  I maintained the structure for compatibility with the intent of improving it later.

Reply:
This looks the same to my eyes.  What other use cases were you thinking about trying to handle here that were currently being handled?